File: clj/014.clj
;; Project Euler -- Problem 14

;; The following iterative sequence is defined for the set of positive integers:

;; n → n/2 (n is even)
;; n → 3n + 1 (n is odd)

;; Using the rule above and starting with 13, we generate the following sequence:
;; 13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1

;; It can be seen that this sequence (starting at 13 and finishing at
;; 1) contains 10 terms. Although it has not been proved yet (Collatz
;; Problem), it is thought that all starting numbers finish at 1.

;; Which starting number, under one million, produces the longest chain?

;; NOTE: Once the chain starts the terms are allowed to go above one million.

;; (defn collatz-sequence
;;   [n]
;;   (if (= n 1)
;;     (seq [1])
;;     (lazy-seq (conj (collatz-sequence (if (even? n) (/ n 2) (inc (* n 3)))) n))))

;; (def m-collatz-sequence (memoize collatz-sequence))

(def collatz-chain-length
  (memoize
   (fn [n]
     (cond (== n 1)   1
           (even? n) (inc (collatz-chain-length (/ n 2)))
           :else     (inc (collatz-chain-length (inc (* 3 n))))))))

(println (first (reduce #(if (> (second %) (second %2)) % %2) (for [x (range 1 1000000)] [x (collatz-chain-length x)]))))

;; (println (ffirst (let [c-map (into {} (for [k (range 1 1000000)] [k (collatz-chain-length k)]))]
;;                     (into (sorted-map-by (fn [k1 k2] (compare [(get c-map k2) k2] [(get c-map k1) k1]))) c-map))))

File: pl/014.pl
#!/usr/bin/perl

# Project Euler -- Problem 14

# The following iterative sequence is defined for the set of positive integers:

# n → n/2 (n is even)
# n → 3n + 1 (n is odd)

# Using the rule above and starting with 13, we generate the following sequence:
# 13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1

# It can be seen that this sequence (starting at 13 and finishing at
# 1) contains 10 terms. Although it has not been proved yet (Collatz
# Problem), it is thought that all starting numbers finish at 1.

# Which starting number, under one million, produces the longest chain?

# NOTE: Once the chain starts the terms are allowed to go above one million.

use 5.10.0;
use strict;
use warnings;
no warnings 'recursion';

use Data::Dump qw< ddx pp >;

my @SEQUENCE_CACHE = ( undef, 1 );

sub is_even { $_[0] % 2 == 0 }

sub collatz_sequence {
    my $n = shift;

    unless ($SEQUENCE_CACHE[$n]) {
        my $tmp_n = $n;
        my $len = 1;
        while ($tmp_n > 1) {
            $tmp_n = is_even($tmp_n) ? $tmp_n / 2 : 3 * $tmp_n + 1;
            if ($SEQUENCE_CACHE[$tmp_n]) {
                $len += $SEQUENCE_CACHE[$tmp_n];
                last;
            }
            else {
                $len++;
            }
        }
        $SEQUENCE_CACHE[$n] = $len;
    }
    return $SEQUENCE_CACHE[$n];
}

my $max_seq_len = 1;
my $max_start = 1;

my $begin = shift || 2;
my $end   = shift || 10;

$end = $begin if $end < $begin;
for ($begin .. $end) {
    my $seq_len = collatz_sequence($_);
    if ($max_seq_len < $seq_len) {
        $max_seq_len = $seq_len;
        $max_start = $_;
    }
    # say STDERR "$_, $seq_len";
}

# pp \@SEQUENCE_CACHE;

say '- ', scalar @SEQUENCE_CACHE;
say "= $max_start, $max_seq_len";

